cpp/include/cudf/detail/copy_if_else.cuh

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/cudf.h>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <utilities/bit_util.cuh>
+#include <utilities/cuda_utils.hpp>
+
+#include <cub/cub.cuh>
+
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+namespace {    // anonymous
+
+template <typename T, typename Filter, bool has_validity>
+__global__
+void copy_if_else_kernel(  column_device_view const lhs,
+                           column_device_view const rhs,
+                           Filter filter,
+                           mutable_column_device_view out,
+                           cudf::size_type * __restrict__ const null_count)
+{   
+   const cudf::size_type tid = threadIdx.x + blockIdx.x * blockDim.x;   
+   const int w_id = tid / warp_size;
+   // begin/end indices for the column data
+   cudf::size_type begin = 0;
+   cudf::size_type end = lhs.size();   
+   // warp indices.  since 1 warp == 32 threads == sizeof(bit_mask_t) * 8,
+   // each warp will process one (32 bit) of the validity mask via 
+   // __ballot_sync()
+   cudf::size_type w_begin = cudf::util::detail::bit_container_index<bit_mask::bit_mask_t>(begin);
+   cudf::size_type w_end = cudf::util::detail::bit_container_index<bit_mask::bit_mask_t>(end);
+
+   // lane id within the current warp
+   const int w_lane_id = threadIdx.x % warp_size;
+
+   // store a null count for each warp in the block
+   constexpr cudf::size_type b_max_warps = 32;
+   __shared__ uint32_t b_warp_null_count[b_max_warps];
+   // initialize count to 0. we have to do this because the WarpReduce
+   // at the end will end up summing all values, even ones which we never 
+   // visit.
+   if(has_validity){   
+      if(threadIdx.x < b_max_warps){
+         b_warp_null_count[threadIdx.x] = 0;
+      }   
+      __syncthreads();   
+   }   
+
+   // current warp.
+   cudf::size_type w_cur = w_begin + w_id;         
+   // process each grid
+   while(w_cur <= w_end){
```

Comment:
```suggestion
   int active_threads = __ballot_sync(0xFFFF'FFFF, w_cur <= w_end);
   while(w_cur <= w_end){
```

Reply:
This loop is setup such that there won't be divergence within a warp, so this isn't technically necessary. It's a bit of a non-traditional way to do a grid-stride loop, but it works.

Reply:
Ahh, I missed that each thread in the warp gets the same begin and end values.